src/Illuminate/Auth/DatabaseUserProvider.php

```diff
@@ -154,8 +154,41 @@ protected function getGenericUser($user)
      */
     public function validateCredentials(UserContract $user, array $credentials)
     {
-        return $this->hasher->check(
-            $credentials['password'], $user->getAuthPassword()
-        );
+        if (is_null($plain = $credentials['password'])) {
+            return false;
+        }
+
+        if (! $this->hasher->check($plain, $hash = $user->getAuthPassword())) {
+            return false;
+        }
+
+        $this->rehashIfRequired($user, $hash, $plain);
+
+        return true;
+    }
+
+    /**
+     * Rehash the user's password if required.
+     *
+     * @param  \Illuminate\Contracts\Auth\Authenticatable  $user
+     * @param  string  $hash
+     * @param  string  $plain
+     * @return void
+     */
+    public function rehashIfRequired(UserContract $user, string $hash, string $plain)
+    {
+        $attribute = method_exists($user, 'getAuthPasswordName') ? $user->getAuthPasswordName() : 'password';
+
+        if (! isset($user->{$attribute}) || ! hash_equals($hash, $user->{$attribute})) {
```

Comment:
Now the hash gets checked twice (once in `validateCredentials` and here).

I guess this is done to be able to make this a public method so it won't rehash passwords that don't match the user's password. But the downside is double hash checking for every login attempt.

Would it make more sense to split this in two methods, one private without the hash check and one public that uses the private method after validating the hash?

Reply:
`hash_equals()` is a basic string comparison function, I could have used `===`, but it's best practice to use that for hashes. 

You're thinking of `password_verify()`.